Storage GC marking encountered zombie ObjectId 1000000000000000030 / No entity found for objectId 1000000000000000030

I found a scenario that triggers a GC warning for a zombie object with a minimal setup.

The issue is that if I try to store a root containing a Lazy into a new storage, I get an exception as soon as I attempt to access that Lazy. However, as long as I don’t touch it, the storage continues to work without problems. Despite that, the logs already contain messages indicating that the GC is removing a zombie object.

So the data behind the Lazy is actually cleaned up by the GC without the user being notified at startup. Later, when the user accesses the Lazy, an exception is thrown.

Overall, this can lead to unnoticed data loss.

**Solution:**

-  either reinitialize the Lazy to the new storage during startup, or

-  immediately abort the startup process (which is the worse option).


```java
package test.microstream.various;

import java.nio.file.Path;

import org.eclipse.serializer.reference.Lazy;
import org.eclipse.store.storage.embedded.types.EmbeddedStorage;
import org.eclipse.store.storage.embedded.types.EmbeddedStorageManager;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

public class TwoStoragesTest
{

	@TempDir
	Path location;


	@Test
	public void saveDefaultLazySecondTimeTest(@TempDir final Path secondLocation)
	{
		final MyRoot myRoot = new MyRoot("Hello World");

		try (EmbeddedStorageManager storageManager = EmbeddedStorage.start(myRoot, this.location)) {
		}

		myRoot.lazy.clear();

		try (EmbeddedStorageManager storageManager = EmbeddedStorage.start(myRoot, secondLocation)) {
			myRoot.lazy.get();
		}

	}


	public static class MyRoot
	{
		Lazy<String> lazy;

		public MyRoot(final String content)
		{
			super();
			this.lazy = Lazy.Reference(content);
		}

	}

}
```

First storage data file:

<img width="351" height="171" alt="Image" src="https://github.com/user-attachments/assets/52cb278f-3ddd-4603-9e32-911384bed466" />


Second storage data file:

<img width="311" height="175" alt="Image" src="https://github.com/user-attachments/assets/6a5185b1-d553-4838-bf5a-640f8adfa9f9" />


debug position:

<img width="891" height="403" alt="Image" src="https://github.com/user-attachments/assets/66d8135e-5fc9-4f93-acc1-37c32de01b1d" />

console:

<img width="1333" height="169" alt="Image" src="https://github.com/user-attachments/assets/44306121-1f0d-4d6c-9f64-b37dc3699543" />

Update: 1
The following scenario:
```java
package test.microstream.various;

import java.nio.file.Path;

import org.eclipse.serializer.reference.Lazy;
import org.eclipse.store.storage.embedded.types.EmbeddedStorage;
import org.eclipse.store.storage.embedded.types.EmbeddedStorageManager;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

public class TwoStoragesTest
{

	@TempDir
	Path location;


	@Test
	public void saveDefaultLazySecondTimeTest(@TempDir final Path secondLocation)
	{
		final MyRoot myRoot = new MyRoot("Hello World");

		try (EmbeddedStorageManager storageManager = EmbeddedStorage.start(myRoot, this.location)) {
		}

		myRoot.lazy.clear();

		try (EmbeddedStorageManager storageManager = EmbeddedStorage.start(myRoot, secondLocation)) {
			System.out.println(myRoot.number);
		}

		try (EmbeddedStorageManager storageManager = EmbeddedStorage.start(secondLocation)) { //<-- Exception
			MyRoot root = (MyRoot) storageManager.root();
			root.lazy.get();
		}

	}


	public static class MyRoot
	{
		Lazy<String> lazy;
		Integer number = 42;

		public MyRoot(final String content)
		{
			super();
			this.lazy = Lazy.Reference(content);
		}

	}

}

```
ends with:
```console
2025-12-17 11:14:23:790 +0100 [Eclipse Serializer-StorageChannel-0] WARN StorageEntityCache - Storage GC marking encountered zombie ObjectId 1000000000000000030
2025-12-17 11:14:23:818 +0100 [Eclipse Serializer-StorageChannel-0] WARN StorageEntityCache - Storage GC marking encountered zombie ObjectId 1000000000000000030
2025-12-17 11:14:23:823 +0100 [Eclipse Serializer-StorageChannel-0] ERROR StorageChannelTask - Error occurred in storage channel#0
org.eclipse.store.storage.exceptions.StorageExceptionConsistency: No entity found for objectId 1000000000000000030
	at org.eclipse.store.storage.types.StorageEntityCollector$EntityCollectorByOid.accept(StorageEntityCollector.java:119)
	at org.eclipse.serializer.persistence.binary.types.LoadItemsChain$ChannelHashing$ChainItemObjectIdSet.iterate(LoadItemsChain.java:324)
	at org.eclipse.store.storage.types.StorageChannel$Default.collectLoadByOids(StorageChannel.java:640)
	at org.eclipse.store.storage.types.StorageRequestTaskLoadByOids$Default.internalProcessBy(StorageRequestTaskLoadByOids.java:52)
	at org.eclipse.store.storage.types.StorageRequestTaskLoadByOids$Default.internalProcessBy(StorageRequestTaskLoadByOids.java:22)
	at org.eclipse.store.storage.types.StorageChannelTask$Abstract.processBy(StorageChannelTask.java:244)
	at org.eclipse.store.storage.types.StorageChannel$Default.work(StorageChannel.java:458)
	at org.eclipse.store.storage.types.StorageChannel$Default.run(StorageChannel.java:541)
	at java.base/java.lang.Thread.run(Thread.java:840)

org.eclipse.store.storage.exceptions.StorageException: Problem in channel #0

	at org.eclipse.store.storage.types.StorageChannelTask$Abstract.checkForProblems(StorageChannelTask.java:106)
	at org.eclipse.store.storage.types.StorageChannelTask$Abstract.waitOnCompletion(StorageChannelTask.java:168)
	at org.eclipse.store.storage.types.StorageRequestAcceptor$Default.waitOnTask(StorageRequestAcceptor.java:165)
	at org.eclipse.store.storage.types.StorageRequestAcceptor$Default.queryByObjectIds(StorageRequestAcceptor.java:194)
	at org.eclipse.store.storage.embedded.types.EmbeddedStorageBinarySource$Default.readByObjectIds(EmbeddedStorageBinarySource.java:98)
	at org.eclipse.serializer.persistence.binary.types.BinaryLoader$Default.readLoadOidData(BinaryLoader.java:775)
	at org.eclipse.serializer.persistence.binary.types.BinaryLoader$Default.getObject(BinaryLoader.java:848)
	at org.eclipse.serializer.persistence.types.PersistenceManager$Default.getObject(PersistenceManager.java:424)
	at org.eclipse.store.storage.types.StorageConnection.getObject(StorageConnection.java:467)
	at org.eclipse.store.storage.types.Database$Default.getObject(Database.java:159)
	at org.eclipse.serializer.reference.Lazy$Default.load(Lazy.java:530)
	at org.eclipse.serializer.reference.Lazy$Default.get(Lazy.java:511)
	at test.microstream.various.TwoStoragesTest.saveDefaultLazySecondTimeTest(TwoStoragesTest.java:34)
Caused by: org.eclipse.store.storage.exceptions.StorageExceptionConsistency: No entity found for objectId 1000000000000000030
	at org.eclipse.store.storage.types.StorageEntityCollector$EntityCollectorByOid.accept(StorageEntityCollector.java:119)
	at org.eclipse.serializer.persistence.binary.types.LoadItemsChain$ChannelHashing$ChainItemObjectIdSet.iterate(LoadItemsChain.java:324)
	at org.eclipse.store.storage.types.StorageChannel$Default.collectLoadByOids(StorageChannel.java:640)
	at org.eclipse.store.storage.types.StorageRequestTaskLoadByOids$Default.internalProcessBy(StorageRequestTaskLoadByOids.java:52)
	at org.eclipse.store.storage.types.StorageRequestTaskLoadByOids$Default.internalProcessBy(StorageRequestTaskLoadByOids.java:22)
	at org.eclipse.store.storage.types.StorageChannelTask$Abstract.processBy(StorageChannelTask.java:244)
	at org.eclipse.store.storage.types.StorageChannel$Default.work(StorageChannel.java:458)
	at org.eclipse.store.storage.types.StorageChannel$Default.run(StorageChannel.java:541)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

<img width="1428" height="413" alt="Image" src="https://github.com/user-attachments/assets/6e58c2c4-6047-4153-a732-8a71a4d4a302" />

--- a/persistence/binary/src/main/java/org/eclipse/serializer/persistence/binary/org/eclipse/serializer/reference/BinaryHandlerLazyDefault.java
+++ b/persistence/binary/src/main/java/org/eclipse/serializer/persistence/binary/org/eclipse/serializer/reference/BinaryHandlerLazyDefault.java
@@ -98,7 +98,7 @@ public final class BinaryHandlerLazyDefault extends AbstractBinaryHandlerCustom<
 		{
 			referenceOid = instance.objectId();
 			
-			if(instance.$getLoader() != handler.getObjectRetriever())
+			if(referenceOid != 0 && instance.$getLoader() != handler.getObjectRetriever())
 			{
 				throw new PersistenceException(
 						"Cannot persist an unloaded lazy reference to another storage. " +

